docs/blog/2019-02-06-government-open-data-site-with-gatsby/index.md

```diff
@@ -0,0 +1,237 @@
+
+We expect we can optimize our code to mitigate our long build times, and having only one build process will help. But we are asking a lot of GraphQL, and we understand the build time is associated with the complexity of our query schema.
+
+### Redundancies
+
+Refactoring the site over time means we often have to manage content in two places. For instance, if we update content in the Jekyll part of the site, but we also have a Gatsby version in the pipeline, we need to update that, too.
+
+In some cases, we've delayed content updates or restyling, acknowledging it makes more sense to do so after the migration of that page or asset.
+
+Thankfully, we have strong communication on our team, and we work in the open, which helps us avoid more extreme problems with redundancy.
+
+### Integrating with Jekyll
+
+Related to redundancy, providing a seamless user experience between the Gatsby and Jekyll portions of the site has been a challenge. There were three issues to resolve:
+
+1. Page URL issue
+2. Gatsby prefetch error
+3. Federalist preview URL
+
+#### Page URL issue
+
+When Gatsby builds the site, it creates a `public` directory for all required site assets. We copied this directory to a `gatsby-public` folder for the Jekyll build to use. However, we didn't want our URLs to include `gatsby-public`, so we needed to add a permalink attribute to our pages' front matter to override the default path.
+
+Fortunately, Gatsby provides a hook into the entire lifecycle of its build process, including an [`onPostBuild` API](https://www.gatsbyjs.org/docs/node-apis/#onPostBuild). We use this hook to add the front matter to our pages and copy the `public` directory to the `gatsby-public` directory.
+
+#### Gatsby prefetch error
+
+Developers tout Gatsby's speed, and [prefetching](https://www.gatsbyjs.org/docs/how-code-splitting-works/) page assets is integral to Gatsby's performance advantages. However, Gatsby isn't aware of our deployment structure, which results in an error when a page loads in the production environment.
+
+Basically, the Jekyll part of the site deploys to a directory that isn't known to Gatsby at build time. Consequently, Gatsby creates a `pages.json` object that contains the wrong locations for files. To deal with this, we use another feature of Gatsby's client API, [`onClientEntry`](https://www.gatsbyjs.org/docs/browser-apis/#onClientEntry). Using `gastby-browser.js`, we override `pages.json` by passing the correct assets to Gatsby.
+
+```javascript
+exports.onClientEntry = () => {
+  // Patch the resource loader
+  const loader = global.___loader
+  if (!loader) return
+
+  let path = window.location.pathname
+
+  if (path.includes("/explore") && usStateIds.includes(statePathId)) {
```

Comment:
Just a note that this snippet is a little confusing. I'm presuming `usStateIds` is somewhere above the scope of this function?

Maybe just hard code an array for now?

Reply:
It's code fenced isn't it? I'm not sure why Jeff chose this particular snippet of code to represent this work around. I'm hesitant to change it, since we published it this way on our blog already.

Reply:
I don't really see what code fencing has to do with this?

I mean that `usStateIds` and `statePathId` are not in the scope of this snippet, so it's harder to follow.

Reply:
I see...I thought you meant the actual logic wouldn't work! 

I guess I could link to the full code here as a compromise?

Reply:
Yeah! That would be great!

The idea comes across, you're basically extending the internal loader to be aware of additional paths, but I can it could be confusing when it's broken out of context like this!

Reply:
I see...that's an array of all 50 states and DC, so it would have taken up some serious screen real estate. I added a link to the repo for that one.